[WIP] Add broadcast_double constructor template

mariaschuld opened the pull request:
**Context:** Part of the base (now "constructor") templates. 

**Description of the Change:**

* Adds ``broadcast_double``, which broadcasts a two-wire unitary over pairs of wires, using the ``even`` flag to determine whether to start with the 0th or 1st wire.

![Screenshot from 2020-02-18 15-02-19](https://user-images.githubusercontent.com/23052011/74738500-b5947e80-525f-11ea-8687-790793fd0aa9.png)

* Adds tests.

* Adds ``broadcast_double`` to the gallery of templates.

![Screenshot from 2020-02-18 15-01-29](https://user-images.githubusercontent.com/23052011/74738502-b6c5ab80-525f-11ea-92f0-c59b56733ce2.png)

mariaschuld commented on tests/test_templates.py:
```diff
@@ -123,7 +124,16 @@
+                               {'parameters': [[1.], [1.]]},
+                               {'block': qml.RX,
+                                'wires': [0, 1]}),
+                              (broadcast_double,
```
Snuck in the Broadcast test here too.

mariaschuld commented on doc/introduction/templates.rst:
```diff
@@ -271,85 +292,4 @@ respective interfaces.
-    ~pennylane.init.strong_ent_layers_normal
-
-:html:`</div>`
+The initialization functions can be found in the User API `~pennylane.init`.
```
Should have done this in a separate PR, but I wanted to delete this unwieldy list since a while and could not resist...

mariaschuld commented on doc/introduction/templates.rst:
```diff
@@ -271,85 +289,4 @@ respective interfaces.
-    ~pennylane.init.strong_ent_layers_normal
-
-:html:`</div>`
+The initialization functions can be found in the API of the :mod:`~.pennylane.init` module.
```
Could not resist to delete those...

mariaschuld commented:
@johannesjmeyer  and @antalszava, there will be a few more PRs to come for the rest of the new templates and it would be great if you can be the designated reviewers for all of those...Thanks!

mariaschuld commented:
Just had a short discussion with @josh146 ... one could very elegantly unite all the constructor/base templates in one function called ``broadcast``, which gets a flag ``structure`` that can be ``None``, ``chain``, ``tree``, ``double``, etc.

josh146 commented:
> Just had a short discussion with @josh146 ... one could very elegantly unite all the constructor/base templates in one function called broadcast, which gets a flag structure that can be None, chain, tree, double, etc.

This is quite elegant I have to say!

mariaschuld commented:
It would require a rewrite of this PR, but is not a major issue...

mariaschuld commented:
And it would avoid code duplication in the docstring and tests. At the moment I cannot think of a reason against it.

The ``broadcast`` function could just sit in the ``templates`` directory, or in ``templates.utils``.

johannesjmeyer commented:
I already mentioned this in the earlier PR, essentially all the new templates amount to repeating a unitary across some wires in a specific way. But I would still make different methods for it and have some underlying method `repeat(template, wires_list)` which handles the application of the template where every method is just responsible for creating the correct wires list. I.e. `broadcast -> [[0], [1], ...]`, `broadcast_double ->[[0,1],[2,3], ...]` etc.

mariaschuld commented:
Can you comment why you'd prefer the different functions? Does that not kill the purpose of limiting code repetition?

josh146 commented:
From a user/UI perspective, I would prefer a single function. mainly for discoverability - all options would appear on a single docstring page, where examples etc. can be shown of all use cases.

mariaschuld commented:
I thought it through a bit longer, and would like to keep separate functions for now. 

The reason is the following: One can also use "random_layer" and "strongly_entangling_layer" as constructors, or "ising" from ``QAOAEmbeddin``. In the end there could potentially be lots of base structures if we take into account hardware architecture. Every base structure could require positional arguments that have no function for the other base templates, such as "even" and "odd" in this PR. They also have different conditions on number of parameters, wires etc. In short - it becomes a giant function with lots of branches, difficult to maintain and keep together.

So I propose to go ahead with this PR and the following base templates for now. Would that be fine @josh146 ?

josh146 commented:
Even though it's a matter of semantics, I don't tend to see `StronglyEntanglingLayers`, `RandomLayers` as template constructors. Partially for what @johannesjmeyer and @co9olguy  mentioned in a previous PR --- these templates have a very well-defined structure, and only a few classical hyperparameters for control.

`qml.broadcast` to me feels a bit different, and much more flexible --- much less structured, it simply takes a unitary and attempts to broadcast it across wires. But yes, purely comes down to semantics and where you want to draw the line between template (specific structure/anstaz, minimal manipulation) and template constructor (takes a template, outputs another template, much more flexible/no constrained ansatz).

antalszava commented on doc/introduction/templates.rst:
```diff
@@ -189,6 +188,25 @@ of other templates.
+Template constructors
+---------------------
+
+Template constructors are a special type of template that takes other templates (or single quantum operations)
```
```suggestion
Template constructors are special types of templates that take other templates (or single quantum operations)
```

antalszava commented on pennylane/templates/constructors/broadcast_double.py:
```diff
@@ -0,0 +1,237 @@
+# Copyright 2018-2019 Xanadu Quantum Technologies Inc.
```
```suggestion
# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```

mariaschuld replied:
Oh wow, sharp eye!

antalszava commented on pennylane/templates/constructors/broadcast_double.py:
```diff
@@ -0,0 +1,237 @@
+def broadcast_double(block, wires, parameters=None, even=True, kwargs={}):
+    r"""Applies a two-qubit unitary to pairs of wires.
+
+    If ``even`` is ``True``, start with the wire pair ``[0, 1]``, else with the pair ``[1, 2]``.
```
This part gives the impression that these lists that are used as wires when applying the blocks (whereas if I understand it correctly, these are *indices*)

antalszava commented on pennylane/templates/constructors/broadcast_double.py:
```diff
@@ -0,0 +1,237 @@
+
+@template
+def broadcast_double(block, wires, parameters=None, even=True, kwargs={}):
+    r"""Applies a two-qubit unitary to pairs of wires.
```
```suggestion
    r"""Applies the same two-qubit unitary ``block`` to pairs of wires.
```

antalszava replied:
It might be worth elaborating slightly more on the logic (however simple it is, when grasped) right at the beginning. Reading for the very first time, it felt slightly as getting into the explanation of the wire-business pretty quickly. :)

This could also be aided by further extending the first sentence (or potentially elaborating on parts of it). The first sentence could also go like "Broadcasts the same two-qubit unitary by applying the same  ``block`` to pairs of wires.". Though these are just ideas for now, not too sure which would be the best to have.

mariaschuld replied:
I tried to implement this advice in the now single ``broadcast`` function.

antalszava commented on pennylane/templates/constructors/broadcast_double.py:
```diff
@@ -0,0 +1,237 @@
+    Args:
+        block (function): quantum gate or template
+        parameters (iterable or None): sequence of parameters for each gate applied
+        even (boolean): if ``True``, pairs are formed starting with the first wire index, else with
```
I'm most probably missing the point here! Since the user is expected to pass the list of wires anyways, could they just not be made to omit the first one if they would like to go with the `even==False` case?

Or if the idea is that most of the times they would go with `range(len(dev.wires))`, we could also document an example with having the counterpart `range(1, len(dev.wires))`.

antalszava commented on pennylane/templates/constructors/broadcast_double.py:
```diff
@@ -0,0 +1,237 @@
+        start_with = 1
+
+    # extract the pairs of wires that the block acts on
+    wire_pairs = [[wires[i], wires[i + 1]] for i in range(start_with, len(wires) - 1, 2)]
```
What happens if we have the same wire passed several times?

mariaschuld replied:
Good point, never thought about this! I think this should be added to the wire test for all templates! Making a note.

antalszava commented on pennylane/templates/constructors/broadcast_double.py:
```diff
@@ -0,0 +1,237 @@
+        parameters (iterable or None): sequence of parameters for each gate applied
+        even (boolean): if ``True``, pairs are formed starting with the first wire index, else with
+                        the second wire index
+        wires (Sequence[int] or int): wire indices that the unitaries act upon
```
It might be worth adding here that `kwargs` would stand for the auxiliary parameters of a `qfunc` in this case (as is in more details in #510 from line 150 in `qnodes/base.py`)

antalszava commented:
Left a couple of comments, having a look only after some delay, sorry about that! Also, feel free to ping me in such cases :).

So far, the idea looks good to me, from a usability point of view I have the following points:

This might be a really nice convenience for users who would not want to have something like the following in their `qfunc`:
```python
for w, p in zip(wire_pairs, parameters):
    block(*p, wires=w, **kwargs)
```
Those, who find that this works for their needs just perfect, might prefer to go with the pythonic way.

I feel that overall broadcasting quantum gates in PennyLane would have the most power, if actually `qml.broadcast` could take any operation, regardless of the number of qubits they act on. It could under the hood call on the single or double (or multiple) broadcast function (that have been created already, which is super nice for separating the logic). Creating an experience for the users that goes something like `qml.broadcast` -> let's try it with multi-qubit operation -> "Oh, it works, how cool!" could create easy workflows and practices for them.

josh146 commented:
> I feel that overall broadcasting quantum gates in PennyLane would have the most power, if actually qml.broadcast could take any operation

@antalszava, theres a mini-ADR available that asks whether there should be a unified function 🙂 @mariaschuld, that's maybe another vote for a single function.

antalszava commented:
Oh, great, slipped my attention, thanks for that! :)

mariaschuld commented:
Decision is to merge all constructors into one function! 

Working on it, so far it looks indeed very elegant...

mariaschuld commented:
@antalszava about your confusion with the odd case - it is for sheer convenience, so the user does not have to wiggle with wire lists, since even & odd layers are rather common.

mariaschuld commented:
I merged the constructors into a function in the ``templates/`` folder. Updated tests, unified docstring and updated ``template.rst``.

I am still getting doc building errors, if anyone knows why please help. (It seems to render fine).

mariaschuld commented:
Make docs errors should be resolved.

mariaschuld commented:
@josh146 , adding you as a reviewer only in case you are curious!

antalszava commented on pennylane/templates/broadcast.py:
```diff
@@ -13,49 +13,59 @@
 @template
-def Broadcast(block, wires, parameters=None, kwargs={}):
-    r"""Applies a (potentially parametrized) single-qubit unitary ``block`` to each wire.
+def broadcast(block, wires, structure="single", parameters=None, kwargs={}):
```
On the first note, I got `ValueError: CNOT: wrong number of wires. 1 wires given, 2 expected.`, when calling `broadcast(qml.CNOT, wires=[0,1,2,3])` which is expected right now.

mariaschuld replied:
About ``structure=None``, I thought about it, but I think it is by no means clear what the default is here. I want to force the user to fix a structure each time. So if you don't mind, I would leave it this way!

mariaschuld replied:
About ``kwargs=None``, thanks, that makes total sense! I didn't know it is even a best practice.

antalszava replied:
Sure! Would you say that the default for `structure` is unclear since in the `double` case there could be `odd` or `even` cases?

mariaschuld replied:
Well, I'd say it is unclear that a user considers ``double`` as a default, and the gate-wire test makes it very verbose... On the other hand, you are right that ``broadcast`` suggests ``single`` and ``double`` as defaults.

We could easily (without feature breaking) introduce a default later, what do you think?

mariaschuld replied:
Sorry, @antalszava, reading through this again your idea makes total sense, since we actually *have* a default `single`, which breaks for two-qubit gates. I understand your point now. Will implement your suggested change!

mariaschuld replied:
The only problem is that templates have no attribute ``num_wires``

mariaschuld replied:
Shall we just make ``structure`` a positional argument?

antalszava replied:
Yes, I think it might be worth making `structure` required without any defaults for now.

Making it `structure=single`  to have a default value seems to imply the following cases:
* The user specifies `structure` (most preferable case, though requires user's attention) :heavy_check_mark: 
* The user does not specify `structure`, but either indeed a single-qubit template was passed as a block and the default value is valid :heavy_check_mark: 
* The user does not specify `structure`, but passes a PennyLane native operation, in which case `structure` can be deduced (e.g. double or so) :heavy_check_mark: 
* The user does not specify `structure` and passes a custom template which is not single-qubit, in which case an error would be raised :heavy_multiplication_x: 

It would be really nice to have some sort of addition perhaps to the `@template` decorator or in other way imposing the constraint onto custom templates that they need to specify the `num_wires`. But this is entirely out of scope for this PR.

antalszava commented on pennylane/templates/__init__.py:
```diff
@@ -15,9 +15,9 @@
 This module contains templates, which are pre-coded routines that can be used in a quantum node.
 """
 
+from .broadcast import *
```
You might also have to modify the `from pennylane.templates import template` line in `pennylane/__init__.py` to `from pennylane.templates import template, broadcast` to make `qml.broadcast` work.

antalszava replied:
Right now `broadcast` needs to be imported

mariaschuld replied:
If I do that we can import ``broadcast`` like ``qml.broadcast``, right? 

I would rather enforce ``qml.templates.broadcast``, so the user knows WHAT gets broadcasted here :)

What do you think?

antalszava replied:
Yep!

Sure, sounds good to me :) Was suggesting `qml.broadcast` as it was mentioned in previous PRs, but `qml.templates.broadcast` is also cool

josh146 replied:
Note that `broadcast` acts on both operations _and_ templates, akin to `qml.inv()` (which I think is also defined in template utils?). So `qml.broadcast` seems more natural to me.

mariaschuld replied:
You two convinced me. Shall we still keep it physically in the templates module, but just add the import structure @antalszava suggested? Or shall I move it into ``pennylane.utils`` ?

josh146 replied:
Where does inv live? It should probably go in the same place.

(Train of thought: I quite dislike utils - it has become a catch all module, and leads to a lot of circular import issues)

mariaschuld replied:
That's bad - ``inv()`` lives in utils :)

We can leave it in ``templates`` for now, since changing the location will not be feature breaking if we allow ``qml.broadcast`` import.

antalszava commented on pennylane/templates/broadcast.py:
```diff
@@ -223,21 +267,43 @@ def circuit(hadamard=None):
 
-    for w, p in zip(wires, parameters):
+    wire_sequence = {
+        "single": wires,
```
Here I still would be in favour for having some sort of check for making sure that the user did not pass the same wire multiple times in `wires`.

An example is as follows, where due to the second `CNOT` being applied, the result changes:
```python
@qml.qnode(dev)  
def circuit():
    qml.Hadamard(0)
    qml.broadcast(qml.CNOT, wires=[1,0,0,1], structure='double')
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
```
This might be something the user doesn't do on purpose (maybe wires is just being passed on).
So a check like
```python
if len(set(wires)) != len(wires):
    #error: some wires were repeated
```
or the like could be included.

antalszava replied:
It would be also worth having a test case for this

mariaschuld replied:
This should be added to ALL templates, which can be done easily - just add to the ``_check_wires`` test. I will make a separate PR if you don't mind, since it applies to all templates.

antalszava replied:
Sure, that sounds really good, thanks! :)

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,243 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
```
```suggestion
# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```

mariaschuld replied:
Sharp eye, thanks!

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,243 @@
+        """Tests that correct gate queue is created when 'block' is a template."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
```
Cool tests here! :)

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,243 @@
+
+
+@template
+def KwargTemplateDouble(par, wires, a=True):
```
Great case! :blush:

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,243 @@
+
+        for idx, gate in enumerate(rec.queue):
+            i = idx % 2
+            assert isinstance(gate, gates[i])
```
It took me some time to parse the logic here, just as a suggestion you could also go with a more verbose version:
```python
first_gate = gates[0]
second_gate = gates[1]
for idx, gate in enumerate(rec.queue):
    if idx % 2 == 0:
        assert isinstance(gate, first_gate)
    else:
        assert isinstance(gate, second_gate)
```

mariaschuld replied:
Thanks, implemented the more readable version!

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,243 @@
+                      parameters=parameters, kwargs={'a': kwarg})
+
+        for gate, target_gate in zip(rec.queue, target_queue):
+            assert isinstance(gate, target_gate)
```
Here checking that the same parameters were also passed might be useful

antalszava replied:
Oh or is this the next test actually? I'm slightly confused here :sweat_smile:

mariaschuld replied:
Yes, there is a separate test for parameters :)

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,243 @@
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(block=RX, wires=[0, 1], parameters=RX)
```
Cool!

antalszava commented on doc/introduction/templates.rst:
```diff
@@ -271,85 +297,5 @@ respective interfaces.
-
-    ~pennylane.init.strong_ent_layers_uniform
-    ~pennylane.init.strong_ent_layers_normal
+The initialization functions can be found in the :mod:`~.pennylane.init` module.
```
One thing that came to my mind here: could you perhaps organize this small change into a separate PR? I'm super happy to approve it right away :blush: 

One of the reasons for this would be cutting down the size of this one, such that more emphasis can be put on viewing and reviewing the changes related to `broadcast`. Also, it would be easy to find this small change in the list of merged PRs. :slightly_smiling_face:

mariaschuld replied:
You are so right :)

josh146 commented:
Regarding the docs check, I deliberately made it treat warnings as errors :) This avoids links breaking, which Sphinx by default will just gives warnings for. In this particular case, because the template decorator is imported into another file, Sphinx isn't sure which one to reference. You could do

```rest
:func:`template decorator <pennylane.template>`
```

to force a location, or even just

```rest
:func:`~.pennylane.template`
```

mariaschuld commented on doc/introduction/templates.rst:
```diff
@@ -189,6 +188,33 @@ of other templates.
+takes other templates (or single quantum operations) and applies them to wires in a specific structure.
+
+.. customgalleryitem::
+    :link: ../code/api/pennylane.templates.broadcast.html
```
Oh I understand now - since I am not importing an automodule, it does not get rendered and putinto the ``api`` folder...Thanks!

mariaschuld commented on pennylane/templates/broadcast.py:
```diff
@@ -13,49 +13,59 @@
+    * ``structure= 'single'`` (the default case) applies a single-wire block to each one of the :math:`M` wires:
 
-    If the block does not depend on parameters, ``parameters`` is set to ``None``.
+      .. figure:: ../_static/templates/broadcast_single.png
```
The reason not to do this previously was that it is actually *not* the same unitary, because it gets different parameters each time it is applied. So the cleaner way would be to use :math:`U(\theta^1_1, \dots \theta^1_D)`, :math:`U(\theta^2_1, \dots \theta^2_D)` ,..., and then things get very messy.

What do you think?

antalszava replied:
That's a fair point, yes! One way would be to simply denote it as `theta` and regard it as a symbolic variable. But then it might actually confuse people into thinking that each time the same arguments are passed to the gate. :thinking: 

Perhaps, after all, it is just best to leave it as is, as you suggested. :)

mariaschuld commented on pennylane/templates/broadcast.py:
```diff
@@ -13,49 +13,59 @@
+      subsequent pairs of wires, starting with the second wire:
 
-        block(parameter1, ... parameterD, wires, **kwargs)
+      .. figure:: ../_static/templates/broadcast_double_odd.png
```
You are right, but maybe 5 wires is better, to show what the even case does with an odd number of wires?

I updates the pics accordingly...

mariaschuld replied:
![Screenshot from 2020-02-24 09-53-17](https://user-images.githubusercontent.com/23052011/75135867-8eb8cb00-56eb-11ea-9531-f922c2588183.png)

mariaschuld commented on pennylane/templates/broadcast.py:
```diff
@@ -140,7 +150,7 @@ def circuit(pars):
             circuit([[1, 1], [2, 1], [0.1, 1]])
 
-        As mentioned above, in general the block **must** have the following signature:
+        In general, the block takes D parameters and **must** have the following signature:
```
But it is not - you could have a block of 100 parameters acting on a single wire: simply define the block as a template that consists of 100 rotations...

Do you think this should be clarified for the user?

antalszava replied:
That's true! What I'm getting at, is that even in that case, I think that you would have the shape as `(1,100)` that is, `(num_wires, num_params_per_wire)`. Would that complicate it too much for the users?

mariaschuld commented:
Thanks so much @antalszava and @josh146, docs are building now and I implemented your changes.

I turned our informal ADR into an official one, and we should wait its approval before merging. However, I wanted to clean up this PR in any case.

antalszava commented on pennylane/templates/embeddings/amplitude.py:
```diff
@@ -27,7 +27,8 @@
 
-TOLERANCE = 1e-3
+# tolerance for normalization
+TOLERANCE = 1e-10
```
This change might be outdated if you merge the `master` branch into `add_base_templates` branch. Since there changes were committed to master, but they were not to `add_base_templates` as per this diff.

antalszava commented on tests/test_templates.py:
```diff
@@ -123,7 +125,17 @@
+                               {'parameters': [[1., 1., 1.]]},
+                               {'block': qml.CRot,
+                                'wires': [0, 1],
+                                'structure': 'double'})
```
It might be worth adding a case where:
* `'block'` is user defined template here
* `'structure'` is `double-odd`

mariaschuld replied:
I was thinking about that and actually tend towards only testing ``broadcast`` with one structure here. These are just integration tests, and the structures themselves working as expected should be fully tested in ``test_templates_broadcast``.

Also, there is a quadratic growth in the number of tests run, since every template gets integrated with every other.

There is a subsequent PR adding more structures where I removed the `double` test case, so if you agree to the above we can leave it as is for now and change it in the next PR.

antalszava replied:
Sounds good! :)

antalszava commented on tests/test_templates.py:
```diff
@@ -624,7 +636,19 @@ class TestGradientIntegration:
                                           {'features': [1., 2.], 'weights': [[0.1, 0.1, 0.1]]},
                                           {'wires': range(2)},
-                                          [1])
+                                          [1]),
```
Same comment as for the previous test case

mariaschuld replied:
See answer above...

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        for gate in rec.queue:
+            assert isinstance(gate, unitary)
```
There could be an assertion here for the parameters as well

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+                                                               ("single", [1, 2, 3], [[1], [2], [3]], RX),
+                                                               ])
+    def test_correct_queue_for_gate_block(self, structure, pars1, pars2, gate):
+        """Tests that specific parameter inputs have the same output."""
```
So just to make sure: this is testing that the same gate can be broadcast by passing arguments to `parameters` that are merely different in structure (e.g. list vs list of lists, list of empty lists vs `None`)?

antalszava replied:
If so, maybe the name could be something like `test_correct_queue_same_gate_block_different_parameter_formats`, as you were using the `_different_parameter_formats` naming later as well :slightly_smiling_face:

mariaschuld replied:
Ah thanks for pointing that out. Having this test for both structures is actually redundant. I renamed as you said and deleted the test for `double`.

antalszava replied:
Oh, actually, in the sense that it tested that the specific block was a `single` block, it was perhaps not entirely redundant. Though hopefully might also not be the most important of tests. :)

antalszava replied:
Also just saw your comment on the integration test cases, that would be another reason to leave it as is

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(block=RX, wires=range(n_wires), parameters=parameters)
```
```suggestion
            broadcast(block=RX, `'structure'=single`, wires=range(n_wires), parameters=parameters)
```

antalszava replied:
Since no default is defined now

co9olguy replied:
Agree with @antalszava, would be best from user's perspective if "single" was the default of `broadcast` (so they'd never have to actually write it!)

antalszava replied:
Yes, that would be really nice @co9olguy as it would provide extra convenience to the user :)

There are certain cases, however, when this would lead to errors. These are the ones where a user defines a multi-qubit template but leaves the default `structure="single"` option. In these cases, we are not too sure how we could deduce on how many qubits the template would act on and would have to rely on the user-specified value.

So the question is whether to:
* cover most cases with a user convenience (by having `"single"` as default), but in certain cases this would result in errors related to wires
* require the user to always pass this as an argument

There was a more detailed discussion in this PR on this issue, though it became quite hidden here on GitHub (might need to wait a bit until the messages open in the thread): https://github.com/XanaduAI/pennylane/pull/515#discussion_r382243785

An alternative way could actually be having a `try-except` block, and catch the first time such an error would be raised. Though this might not be too elegant of a solution.

co9olguy replied:
Might be enough to just raise an error if the default is used with a template of a size that doesn't naturally make sense

antalszava replied:
Yes, that's a good idea! The thing is, that for templates it's not so easy how to deduce the size (but maybe we have missed something in our discussion). For PL `Operations` it is contained in `num_wires`, but user-defined templates are simple callable functions.

One way would be to modify the `Recorder` class. Since in its `queue` the operation instances (including the `wires` they are called with) are already present, the overall `wires` could be extracted for the template as well. A question I'm unsure how to solve is where to store this information such that it can be extracted easily (the `template` decorater which uses the `Recorder` class merely returns `recorder.queue`, a list of operations). All in all, this approach would have a couple of edge cases in line and might actually require more thinking.

josh146 replied:
Could the template decorator be modified to add attributes to the python function? e.g.,

```python
@template
def ansatz(weights):
    qml.template.StronglyEntanglingLayers(weights, wires=[0, 2, 3])
    qml.Hadamard(wires=4)

>>> ansatz.num_wires
4
>>> anstaz.wires
[0, 2, 3, 4]
```

mariaschuld replied:
That would be cool @josh146 .

Thanks for summarising all the previous discussion so well @antalszava. I propose we run with no default for now and reassess this, since a) it will not be feature breaking and b) it is really a lot more complex.

josh146 replied:
> That would be cool @josh146 .

I believe it should be possible, we can create an issue for it for a future PR.

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(block=RX, wires='a', parameters=[[1], [2]])
```
```suggestion
            broadcast(block=RX, wires='a', structure='single', parameters=[[1], [2]])
```

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(block=RX, wires=[0, 1], parameters=RX)
```
```suggestion
            broadcast(block=RX, wires=[0, 1], structure='single', parameters=RX)
```

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+            circuit()
+
+    def test_exception_default_case_not_valid(self):
+        """Tests that an exception is raised if 'structure=None' for a 'block' acting on more than
```
This test could be modified such that there is a difference between the template that was specified and the `structure` passed, that is e.g. `structure='double'`, but `block=ThreeWireTemplate`.

mariaschuld replied:
Sorry, I left halfway in between editing this PR and hadn't updated the tests yet. 

For now I just delete the test, since we go with "no default". Again, for this to work elegantly we'd need a wire checking function for templates.

antalszava commented on tests/test_templates_broadcast.py:
```diff
@@ -0,0 +1,267 @@
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(block=ThreeWireTemplate, wires=range(3))
```
```suggestion
            broadcast(block=ThreeWireTemplate, wires=range(3), structure="double")
```

co9olguy commented:
I mentioned elsewhere, but I suggest `pattern` instead of `structure`. I think it is more intuitive when combined with the term `broadcast`

mariaschuld commented:
> I mentioned elsewhere, but I suggest `pattern` instead of `structure`. I think it is more intuitive when combined with the term `broadcast`

I like that. Refactored the argument.

On this note, I am also heavily in favour of changing ``block`` to ``unitary`` (this was the original name), but got voted out. Opportunity to reassess?

mariaschuld commented:
Ready for final review. Thanks everyone!

antalszava commented on doc/introduction/templates.rst:
```diff
@@ -189,6 +187,33 @@ of other templates.
+---------------------
+
+PennyLane offers a broadcasting function to easily construct templates. :func:`~.broadcast`
+takes other templates (or single quantum operations) and applies them to wires in a specific structure.
```
```suggestion
takes other templates (or single quantum operations) and applies them to wires in a specific pattern.
```

antalszava replied:
Or a comparable wording, this minor suggestion is mainly aimed at using the final chosen word from the signature of `broadcast`

co9olguy commented:
> On this note, I am also heavily in favour of changing `block` to `unitary` (this was the original name), but got voted out. Opportunity to reassess?

Sure, we just have to finalize UI/naming before final merging/release, but can be done in other PRs
